File: benchmark.cpp
#include "crossmods.hpp"
#include <iostream>
#include <nlohmann/json.hpp>
#include <chrono>
#include <omp.h>
typedef std::chrono::high_resolution_clock Clock;

using json = nlohmann::json;
using namespace crossmods;

int main() {
	double dt = 1/30.0;
	Vddm vddm = {
		.dt=dt,
		.std=0.75,
		.damping=1.6,
		.tau_threshold=2.3,
		.scale=1.0,
	};

	LognormalTdm tdm(1.0, 1.0, 1.0, 1.0, 0.0);
	
	size_t dur = 10.0/dt;
	Grid1d grid({-3.0, 3.0, 100});
	CrossingPdf pdf(dur, dt);
	
	double tau0 = 3.0;
	//double speed = 20.0;
	vector<double> tau(dur);
	std::generate(tau.begin(), tau.end(),
			[&, t=0]() mutable { return tau0 - t++*dt;}
			);
	
	tdm.decisions(tau, dt);

	size_t niter = 10;
	size_t ntrials = 6;
	// OMG C++!!
	vddm.decisions(grid, pdf, tau);
	
	auto t0 = Clock::now();
	for(size_t i=0; i < niter; ++i) {
		for(size_t t=0; t < ntrials; ++t) {
			CrossingPdf pdf(dur, dt);
			vddm.decisions(grid, pdf, tau);
		}
	}

	std::chrono::duration<double, std::milli> fuckyou(Clock::now() - t0);
	std::cerr << fuckyou.count()/(niter*ntrials) << std::endl;
	std::cout << json(pdf.ps) << std::endl;
}

File: benchmark.py
import crossmods
import numpy as np
import matplotlib.pyplot as plt
import time

def main():
    dt = 1/30.0
    vddm = crossmods.Vddm(
            dt=dt, std=0.75,
            damping=1.6,
            tau_threshold=2.3,
            pass_threshold=0.0,
            )
    
    dur = 20.0
    tau0 = 10.0

    dur = tau0
    ts = np.arange(0, dur, dt)
    tau = tau0 - ts
    
    tau_block = tau - 3.0

    grid = crossmods.Grid1d(-3.0, 3.0, 100)
    
    pdf = vddm.blocker_decisions(grid, tau, tau_block)
    #pdf = vddm.decisions(grid, tau)
    samplesize = 1000
    sample = np.random.choice(ts, size=samplesize, p=np.array(pdf.ps)/np.sum(pdf.ps))
    sample[np.random.rand(samplesize) < pdf.uncrossed] = np.inf
    
    #plt.hist(sample[np.isfinite(sample)], bins=np.arange(ts[0], ts[-1], 0.2), density=True)
    #plt.plot(ts, np.array(pdf.ps)/(1 - pdf.uncrossed))
    #plt.show()

    #print(list(map(pdf, sample)))
    niter = 100
    

    start = time.perf_counter()
    for i in range(niter):
        #loglik = vddm.loglikelihood(grid, tau, sample)
        #loglik = vddm.decisions(grid, tau).loglikelihood(sample)
        loglik = vddm.blocker_decisions(grid, tau, tau_block).loglikelihood(sample)
    
    print((time.perf_counter() - start)/niter*1000)
    plt.plot(ts, pdf.ps)
    plt.twinx()
    plt.plot(ts, tau)
    plt.axvline(ts[np.argmin(np.abs(tau_block))], color='black')
    plt.show()


if __name__ == '__main__': main()
